src/main/java/com/atlassian/db/replica/api/DualConnection.java

```diff
@@ -50,34 +27,49 @@
     private final ReplicaConnectionProvider connectionProvider;
     private final ReplicaConsistency consistency;
     private final DatabaseCall databaseCall;
+    private final Set<String> readOnlyFunctions;
 
     private DualConnection(
         ConnectionProvider connectionProvider,
         ReplicaConsistency consistency,
         DatabaseCall databaseCall,
-        StateListener stateListener
+        StateListener stateListener,
+        Set<String> readOnlyFunctions
     ) {
         this.connectionProvider = new ReplicaConnectionProvider(connectionProvider, consistency, stateListener);
         this.consistency = consistency;
         this.databaseCall = databaseCall;
+        this.readOnlyFunctions = readOnlyFunctions;
     }
 
     @Override
     public Statement createStatement() throws SQLException {
         checkClosed();
-        return ReplicaStatement.builder(connectionProvider, consistency, databaseCall).build();
+        return ReplicaStatement.builder(connectionProvider, consistency, databaseCall, readOnlyFunctions).build();
     }
 
     @Override
     public PreparedStatement prepareStatement(String sql) throws SQLException {
         checkClosed();
-        return new ReplicaPreparedStatement.Builder(connectionProvider, consistency, databaseCall, sql).build();
+        return new ReplicaPreparedStatement.Builder(
+            connectionProvider,
+            consistency,
+            databaseCall,
+            sql,
+            readOnlyFunctions
+        ).build();
     }
 
     @Override
     public CallableStatement prepareCall(String sql) throws SQLException {
         checkClosed();
-        return new ReplicaCallableStatement.Builder(connectionProvider, consistency, databaseCall, sql).build();
+        return new ReplicaCallableStatement.Builder(
+            connectionProvider,
+            consistency,
+            databaseCall,
+            sql,
+            readOnlyFunctions
```

Comment:
This should be optional (empty list is a good default), so it shouldn't be required by the builder constructor

Reply:
I want to always propagate `readOnlyFunctions` from DualConnectin to Statements. Not propagating it is a bug. I don't want it to have its own default. It's not API. It's internal.